src/Sarif/Map/JsonMapBuilder.cs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Map
+{
+    /// <summary>
+    ///  JsonMapBuilder constructs a JsonMap for input Json files. It is passed
+    ///  a size percentage target which controls how much detail it includes.
+    /// </summary>
+    public class JsonMapBuilder
+    {
+        private const long NodeSizeEstimateBytes = 90;         // Bytes for one Node with containing property but not including children
+        private const long ArrayStartSizeEstimateBytes = 5;    // Bytes for one ArrayStart location, if under 10KB.
+
+        public double MaxFileSizePercentage { get; private set; }
+        public int MinimumSizeForNode { get; private set; }
+
+        /// <summary>
+        ///  Construct a JsonMapBuilder for the given target output size.
+        ///  (ex: 0.01 means 1% of the input file size)
+        /// </summary>
+        /// <param name="maxFileSizePercentage">Maximum Size of the Map relative to the source file (0.10 means 10%)</param>
+        public JsonMapBuilder(double maxFileSizePercentage)
```

Comment:
>JsonMapBuilder [](start = 15, length = 14)

probably want to ensure maxFilePercentage range is good here?

Reply:
Added verification it's > 0 and < 100. I can't say that values within that range aren't potentially useful. I use 20x in some tests and am trying 0.0001 for the huge file.